File: backend/app/reconciliation/buckets.py
"""Single source of truth for ITC bucket assignment.

A line item ends up in one of three buckets, in priority order:

  1. The CA override (if set) — `approved` → safe, `disputed` → at_risk,
     `ignore` → excluded from totals.
  2. Otherwise, the engine's match_status — `exact`/`fuzzy` → safe,
     `partial`/`no_match` → at_risk.

The frontend mirrors this exact logic in
`frontend/src/lib/reconciliation/buckets.ts` so the badge a CA sees in the
table is the bucket the row will actually contribute to. Keep both in sync.
"""
from __future__ import annotations

from typing import Literal, Optional

from .schemas import CAOverride, MatchStatus

Bucket = Literal["safe", "at_risk", "ignored"]

_SAFE_MATCHES = {MatchStatus.EXACT.value, MatchStatus.FUZZY.value}


def effective_bucket(
    match_status: str, ca_override: Optional[str],
) -> Bucket:
    """Returns the bucket this row contributes to in the headline KPI totals.

    Accepts raw string values (as stored in `recon_line_items`) to avoid
    needing the caller to import the enum.
    """
    if ca_override == CAOverride.APPROVED.value:
        return "safe"
    if ca_override == CAOverride.DISPUTED.value:
        return "at_risk"
    if ca_override == CAOverride.IGNORE.value:
        return "ignored"
    return "safe" if match_status in _SAFE_MATCHES else "at_risk"

File: frontend/src/lib/reconciliation/buckets.ts
/**
 * Mirror of `backend/app/reconciliation/buckets.py`.
 *
 * Single source of truth for which ITC bucket a recon line item contributes
 * to in the headline KPI totals. The frontend uses this for the per-row
 * badge in the table and the live preview in the drawer; the backend uses
 * the same logic when re-aggregating after a CA override.
 *
 * Priority:
 *   1. If `ca_override` is set, it wins:
 *      - "approved"  → "safe"
 *      - "disputed"  → "at_risk"
 *      - "ignore"    → "ignored" (excluded from both buckets)
 *   2. Otherwise, fall back to match_status:
 *      - exact / fuzzy   → "safe"
 *      - partial / no_match → "at_risk"
 */
import type { CAOverride, MatchStatus, ReconLineItemRow } from "@/types/reconciliation"

export type Bucket = "safe" | "at_risk" | "ignored"

const SAFE_MATCHES = new Set<MatchStatus>(["exact", "fuzzy"])

export function effectiveBucket(
  matchStatus: MatchStatus,
  caOverride: CAOverride | null | undefined,
): Bucket {
  if (caOverride === "approved") return "safe"
  if (caOverride === "disputed") return "at_risk"
  if (caOverride === "ignore") return "ignored"
  return SAFE_MATCHES.has(matchStatus) ? "safe" : "at_risk"
}

/** Convenience for the common "I have a row, give me its bucket" call site. */
export function rowBucket(item: ReconLineItemRow): Bucket {
  return effectiveBucket(item.match_status, item.ca_override)
}

export const BUCKET_LABELS: Record<Bucket, string> = {
  safe: "Safe ITC",
  at_risk: "At-risk ITC",
  ignored: "Ignored",
}

/**
 * Human-readable explanation for a row's bucket assignment. Used by the
 * drawer's live-preview pill so the CA understands why the row is in the
 * bucket it's in (and what changes if they pick a different override).
 */
export function bucketReason(
  matchStatus: MatchStatus,
  caOverride: CAOverride | null | undefined,
): string {
  if (caOverride === "approved") return "You approved this row, so it counts as Safe ITC."
  if (caOverride === "disputed") return "You disputed this row, so it counts as At-risk ITC."
  if (caOverride === "ignore")   return "You ignored this row, so it is excluded from the totals."
  if (matchStatus === "exact")    return "Exact match — counts as Safe ITC by default."
  if (matchStatus === "fuzzy")    return "Fuzzy match (BIN+invoice agree, minor date/amount drift) — counts as Safe ITC by default."
  if (matchStatus === "partial")  return "Partial match (BIN found in supplier export but invoice number doesn't line up) — counts as At-risk ITC by default."
  return "No match in the supplier-filed export — counts as At-risk ITC by default."
}
